File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/JavaScript/LRUcache.js

//Design and implement a data structure for Least Recently Used (LRU) cache. It should support the following operations: get and set.
//
//get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
//set(key, value) - Set or insert the value if the key is not already present. When the cache reached its capacity, it should invalidate the least recently used item before inserting a new item.


var key_value_map = {};         // object used to store each keys' array position
var key_array = [];     // used to store values, the one at the end is the least recently used one
var size = 0, cache_capacity = 0;
/**
 * @constructor
 */
var LRUCache = function(capacity) {
  key_value_map = {};
  key_array = [];
  size = 0;
  cache_capacity = 0;
  cache_capacity = capacity;
};

/**
 * @param {number} key
 * @returns {number}
 */
LRUCache.prototype.get = function(key) {
  //update position if necessary
  if(key_value_map[key]) {
    key_array.splice(key_array.indexOf(key), 1)
    key_array.unshift(key);
    return key_value_map[key];
  } else {
    return -1;
  }
};

/**
 * @param {number} key
 * @param {number} value
 * @returns {void}
 */
LRUCache.prototype.set = function(key, value) {
  if(key_value_map[key]) {
    key_array.splice(key_array.indexOf(key), 1)
    key_array.unshift(key);
    key_value_map[key] = value; // update value if necessary
    return;
  }

  // Not previously exist
  if(size === cache_capacity) {
    key_array.unshift(key);
    key_value_map[key] = value;
    delete key_value_map[(key_array.pop())];
  } else {
    size++;
    key_value_map[key] = value;
    key_array.unshift(key);
  }
};


var testCache = new LRUCache(4);
console.log(testCache.set(1,1));
console.log(testCache.set(2,2));
console.log(testCache.set(3,3));
console.log(testCache.set(4,4));
console.log(testCache.set(5,5));
console.log(testCache.set(6,6));

console.log(testCache.get(1));
console.log(testCache.get(2));
console.log(testCache.get(3));
console.log(testCache.get(4));
console.log(testCache.get(5));
console.log(testCache.get(6));

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/LRUCache.py
class LRUCache:
    """ Implement Least Recent Used Cache"""    
    # @param capacity, an integer
    def __init__(self, capacity):
        """constructor, initialize capacity"""
        self._used_list_ = list()   # with size capacity
        self._cache_ = dict()
        self._size_ = 0
        self._capacity_ = capacity 
    
    # @return an integer
    def get(self, key):
        """ return the value of key"""
    
        # if key in dict, return dict[key], otherwise -1
        if key in self._cache_:
            # update the key to most recent used one
            self._used_list_.remove(key)
            self._used_list_.insert(0,key)
            return self._cache_[key]
        else:
            return -1

    
    # @param key, an integer
    # @param value, an integer
    # @return nothing
    def set(self, key, value):
        """ 
        Set the value for key, if cache is full
        removed the least recent used one
        """
        
        # check if it's already in, if so, update to most recent used
        if key in self._cache_:
            self._cache_[key] = value       # udpate value
            self._used_list_.remove(key)
            self._used_list_.insert(0, key)
        else:
            if self._size_ == self._capacity_:
                del self._cache_[self._used_list_[-1]]        
                self._used_list_.pop()      
                self._used_list_.insert(0, key)
                self._cache_[key] = value
            else:
                self._used_list_.insert(0, key)
                self._cache_[key] = value
                self._size_ += 1
        
if __name__ == "__main__":
    inst = LRUCache(10)
    inst.set(1 ,1 )
    inst.set(2 ,2 )
    inst.set(3 ,3 )
    inst.set(4 ,4 )
    inst.set(5 ,5 )
    inst.set(6 ,6 )
    inst.set(7 ,7 )
    inst.set(8 ,8 )
    inst.set(9 ,9 )
    inst.set(10,10)
    inst.set(11,11)
    inst.set(12,12)

    print inst.get(1)
    print inst.get(2)
    print inst.get(3)
    print inst.get(4)
    print inst.get(5)
    print inst.get(6)
    print inst.get(7)
    print inst.get(8)
    print inst.get(9)
    print inst.get(10)
    print inst.get(11)
    print inst.get(12)
      
